ffi/src/log.rs

```diff
@@ -0,0 +1,187 @@
+use std::{
+    fmt::{Debug, Formatter},
+    mem::ManuallyDrop,
+    sync::Once,
+};
+
+pub use log::LevelFilter;
+use log::{Level, Log, Metadata, Record};
+use simple_logger::SimpleLogger;
+
+/// Configure logging for the library.
+#[no_mangle]
+pub extern "C" fn set_log_level(level: LevelFilter) -> i32 {
+    match SimpleLogger::new().with_level(level).init() {
+        Ok(_) => 0,
+        Err(e) => {
+            eprintln!("Failed to initialize logging: {e:?}");
+            -1
+        }
+    }
+}
+
+/// FFI-safe borrowed Rust &str. Can represent `Option<&str>` by setting ptr to null.
+#[repr(C)]
+pub struct CStr {
```

Comment:
Can we try this type? If I understood it correctly it will result in `char *` on the `ffi` side which is easier to use on the Go side than a custom type.

Reply:
No, it's not FFI-safe and it's not meant to be passed between C and Rust. That type is used in Rust to convert between C strings (`*const c_char`) and Rust equivalents (`String` and `&str`). The recommended way is to pass strings as `*const c_char`.

Reply:
I still think that `std::ffi::CString` is the right call here and not a custom type. The documentation for it suggests to use it to convert between Rust strings and C strings and it can be used to pass them as `*const c_char`? It's also much more common to use `\0` terminated strings in C than having to pass lengths along a pointer 🤷 .

This is also the first thing that pops up when googling for "convert string Rust C": https://stackoverflow.com/questions/24145823/how-do-i-convert-a-c-string-into-a-rust-string-and-back-via-ffi along with https://doc.rust-lang.org/std/ffi/struct.CString.html